### PR TITLE
Add release Dockerfiles

### PR DESCRIPTION
**Pull request by alanking**
This approach scales better and provides a little more flexibility than the previous iteration (see #114).

We may wish to discuss default behaviors and how these options are specified.

I am also a little unsure about the use of environment variables in this way for the build-args, but it makes things a lot more flexible this way.

With all this in mind, leaving in draft again :)

### CODE REVIEW

File: irods_testing_environment/database_setup.py [hunk 1]
```diff
@@ -378,3 +378,43 @@ def setup_catalog(ctx,
         raise RuntimeError('failed to grant privileges to user [{0}] on database [{1}]'.format(database_user, database_name))
 
     strat.list_databases()
+
+def wait_for_database_service(ctx,
```
Comment (alanking) [inline irods_testing_environment/database_setup.py:382]: Inspired by https://github.com/korydraughn/irods_docker/blob/b027f0e453ff4f0f9afcd3e7d1a3b13eabb54f2a/compose/just_stand_it_up/ubuntu_20/mysql/provider/wait_for_database.py

File: irods_testing_environment/database_setup.py [hunk 2]
```diff
@@ -378,3 +378,43 @@ def setup_catalog(ctx,
...
+                logging.debug(
+                    f'trying database on [{db_container.name}] ip:[{db_address}] port:[{db_port}]')
+                s.connect((db_address, db_port))
+                return
+
+            except ConnectionRefusedError:
+                retries = retries + 1
+                time.sleep(seconds_between_retries)
+
+        logging.error(f'maximum retries reached attempting to connect to database [{db_container.name}]')
```
Comment (alanking) [inline irods_testing_environment/database_setup.py:420]: We should probably raise a `RuntimeError` here, whoops

Comment (alanking) [reply irods_testing_environment/database_setup.py]: Done

**Comment by korydraughn** [comment]
I don't see a problem with using environment variables. My questions are:
- Is the `dockerfile` environment variable safe from the user's environment (i.e. collisions)?
- What is the alternative to using environment variables?
- What default behavior are you referring to?

**Comment by alanking** [comment]
>     * Is the `dockerfile` environment variable safe from the user's environment (i.e. collisions)?
If the user has defined `dockerfile` or `irods_package_version` in their environment and uses the `--use-static-image` option, they will be ignored because the options set the environment variables before invoking `up`. This means that `--use-static-image` is basically a convenience for the user.

Having the environment variables set manually and using other options could lead to unexpected results. For instance, it may be possible to bypass using some options by setting the environment variables and this could allow for using incompatible parameters and triggering unexplored code paths.

I'm open to ideas about what this means and how much to protect the user from this, if at all.

>     * What is the alternative to using environment variables?
We could use the Compose python API to define a `dockerfile` and the `build-arg`s. This makes me uneasy as the Compose python API usage in this project is unsupported and never has been. According to the Compose maintainers, the python API has never been supported as such and enforced this by dropping it entirely in V2 (see https://github.com/irods/irods_testing_environment/issues/8#issuecomment-1021320731).

We could also define separate Compose projects for the release images on each platform, but that seemed even less palatable.

The environment variables were inspired by other Compose projects I've worked with (namely, ZMT).

I'm open to implementing these or any other ideas.

>     * What default behavior are you referring to?

I was mainly thinking about what our expectations are around what happens when some options are used and others are not.

For instance, as we discussed "offline" this morning, does it make sense to require `--irods-package-version` with `--use-static-image` or should we provide a strong default for `--irods-package-version` so that it's not required?

Or, should `--use-static-image` be the default in some cases (e.g. `run_plugin_tests.py`) and we provide a different option to download and install packages at runtime (e.g. `--use-dynamic-image`)?

**Comment by korydraughn** [comment]
I don't think we have to worry about the environment variables. Adding a few statements about those would be good. If we tell the reader that `dockerfile` and `irods_package_version` should not be defined in their testing environment, then I think we're covered.

As for the alternative (i.e. compose api), I agree that attempting to use that api is not in our best interest given it has been dropped in V2.

As for defaults, requiring users to pass `--use-static-image` feels correct. Developers are less likely to run the wrong version of iRODS that way. If a user provides the static image option, the package version should default to the latest release for that platform IMO. The idea that a user should opt into a static image also fits the purpose of this repo too.

**Comment by alanking** [comment]
Okay, so, I think the next steps include the following...

- [ ] Add words to README regarding environment variables
- [ ] Make `--irods-package-version` optional with `--use-static-image` by providing a strong default of "latest release"

**Comment by alanking** [comment]
I created #121 to address the bit about the default behavior of `--irods-package-version`. This is going to be more of a lift because even if you don't specify `irods_package_version` environment variable, a blank string will be used instead which is obviously not good.

File: README.md [hunk 3]
```diff
@@ -154,6 +155,17 @@ Note: Compose project names default to the name of the directory which contains
 
 If all else fails, this script includes a `--help` option which explains what each of the options do and how they are supposed to be used.
 
+## Using static images with released iRODS versions
+
+`--use-static-images` enables building and running a Docker image with iRODS packages installed rather than downloading an installing the packages at runtime. This is useful for testing plugins against released versions of iRODS or for quickly reproducing issues on past versions.
```
Comment (trel) [inline README.md:160]: "downloading an installing" -> "downloading and installing"

Comment (alanking) [reply README.md]: Done

Comment (alanking) [reply README.md]: Done

**Comment by korydraughn** [comment]
This seems like it's in a good spot.

Anything left to do for this?

**Comment by alanking** [comment]
Everything else left to do has been captured in an issue, I believe. I think it's good to go, too

**Pull request merged**